README.md
[![Build Status](https://travis-ci.org/karupanerura/DateTimeX-Moment.svg?branch=master)](https://travis-ci.org/karupanerura/DateTimeX-Moment)
# NAME

DateTimeX::Moment - DateTime like interface for Time::Moment

# SYNOPSIS

    use DateTimeX::Moment;

    $dt = DateTimeX::Moment->new(
        year       => 1964,
        month      => 10,
        day        => 16,
        hour       => 16,
        minute     => 12,
        second     => 47,
        nanosecond => 500000000,
        time_zone  => 'Asia/Taipei',
    );

    $dt = DateTimeX::Moment->from_epoch( epoch => $epoch );
    $dt = DateTimeX::Moment->now; # same as ( epoch => time() )

    $year   = $dt->year;
    $month  = $dt->month;          # 1-12

    $day    = $dt->day;            # 1-31

    $dow    = $dt->day_of_week;    # 1-7 (Monday is 1)

    $hour   = $dt->hour;           # 0-23
    $minute = $dt->minute;         # 0-59

    $second = $dt->second;         # 0-61 (leap seconds!)

    $doy    = $dt->day_of_year;    # 1-366 (leap years)

    $doq    = $dt->day_of_quarter; # 1..

    $qtr    = $dt->quarter;        # 1-4

    # all of the start-at-1 methods above have corresponding start-at-0
    # methods, such as $dt->day_of_month_0, $dt->month_0 and so on

    $ymd    = $dt->ymd;           # 2002-12-06
    $ymd    = $dt->ymd('/');      # 2002/12/06

    $mdy    = $dt->mdy;           # 12-06-2002
    $mdy    = $dt->mdy('/');      # 12/06/2002

    $dmy    = $dt->dmy;           # 06-12-2002
    $dmy    = $dt->dmy('/');      # 06/12/2002

    $hms    = $dt->hms;           # 14:02:29
    $hms    = $dt->hms('!');      # 14!02!29

    $is_leap  = $dt->is_leap_year;

    # these are localizable, see Locales section
    $month_name  = $dt->month_name; # January, February, ...
    $month_abbr  = $dt->month_abbr; # Jan, Feb, ...
    $day_name    = $dt->day_name;   # Monday, Tuesday, ...
    $day_abbr    = $dt->day_abbr;   # Mon, Tue, ...

    # May not work for all possible datetime, see the docs on this
    # method for more details.
    $epoch_time  = $dt->epoch;

    $rhs = $dt + $duration_object;

    $dt3 = $dt - $duration_object;

    $duration_object = $dt - $rhs;

    $dt->set( year => 1882 );

    $dt->set_time_zone( 'America/Chicago' );

    $dt->set_formatter( $formatter );

# BENCHMARK

`author/benchmark.pl`:

    new()
    Benchmark: timing 100000 iterations of datetime, moment...
      datetime:  4 wallclock secs ( 4.06 usr +  0.01 sys =  4.07 CPU) @ 24570.02/s (n=100000)
        moment:  1 wallclock secs ( 0.62 usr +  0.01 sys =  0.63 CPU) @ 158730.16/s (n=100000)
                 Rate datetime   moment
    datetime  24570/s       --     -85%
    moment   158730/s     546%       --
    ----------------------------------------
    now()
    Benchmark: timing 100000 iterations of datetime, moment...
      datetime:  4 wallclock secs ( 4.38 usr +  0.01 sys =  4.39 CPU) @ 22779.04/s (n=100000)
        moment:  1 wallclock secs ( 0.59 usr +  0.00 sys =  0.59 CPU) @ 169491.53/s (n=100000)
                 Rate datetime   moment
    datetime  22779/s       --     -87%
    moment   169492/s     644%       --
    ----------------------------------------
    from_epoch()
    Benchmark: timing 100000 iterations of datetime, moment...
      datetime:  4 wallclock secs ( 4.27 usr +  0.01 sys =  4.28 CPU) @ 23364.49/s (n=100000)
        moment:  1 wallclock secs ( 0.63 usr +  0.00 sys =  0.63 CPU) @ 158730.16/s (n=100000)
                 Rate datetime   moment
    datetime  23364/s       --     -85%
    moment   158730/s     579%       --
    ----------------------------------------
    calculate()
    Benchmark: timing 100000 iterations of datetime, moment...
      datetime: 20 wallclock secs (20.30 usr +  0.04 sys = 20.34 CPU) @ 4916.42/s (n=100000)
        moment:  1 wallclock secs ( 1.07 usr +  0.00 sys =  1.07 CPU) @ 93457.94/s (n=100000)
                Rate datetime   moment
    datetime  4916/s       --     -95%
    moment   93458/s    1801%       --
    ----------------------------------------

# DESCRIPTION

TODO: write it

# METHODS

TODO: write it

# LICENSE

Copyright (C) karupanerura.

This is free software, licensed under:
  The Artistic License 2.0 (GPL Compatible)

# AUTHOR

karupanerura &lt;karupa@cpan.org>


Changes
Revision history for Perl extension DateTimeX-Moment

{{$NEXT}}

0.06 2019-06-21T21:41:01Z

    - 19c8238 add dependency for DateTime::Duration (fix #3)
    - 1a6b3b2 Align/Group accessors and avoid shifting the stack (chansen)
    - e28422d Group methods in a somewhat logical order (chansen)
    - 85a5436 Removed unused routine _mod_and_keep_sign (chansen)
    - 6ea316a Fixed ->delta_days (chansen)
    - f9eb6fb Fixed ->delta_md (chansen)
    - 8ded533 Fixed ->subtract_datetime_absolute (chansen)
    - eab6307 Optimize ->set (chansen)

0.05 2016-04-13T03:39:17Z

    - more increase compatibility with DateTime (thanks chansen)

0.04 2016-04-07T02:12:01Z

    - increase compatibility with DateTime (thanks chansen)

0.03 2016-03-18T10:45:30Z

    - no require DateTime/DateTime::Duration for testing.

0.02 2016-02-15T06:47:15Z

    - fixed deps

0.01 2016-02-13T00:36:17Z

    - original version



lib/DateTime/Moment/Duration.pm
package DateTime::Moment::Duration;
use strict;
use warnings;

use Carp;
use List::Util qw/first/;
use Scalar::Util qw/blessed/;

use constant ALL_UNITS => qw/months days minutes seconds nanoseconds/;

use overload (
    fallback => 1,
    '+'      => '_add_overload',
    '-'      => '_subtract_overload',
    '*'      => '_multiply_overload',
    '<=>'    => '_compare_overload',
    'cmp'    => '_compare_overload',
);

sub isa {
    my ($invocant, $a) = @_;
    return !!1 if $a eq 'DateTime::Duration';
    return $invocant->SUPER::isa($a);
}

sub new {
    my $class = shift;
    my %args = (@_ == 1 && ref $_[0] eq 'HASH') ? %{$_[0]} : @_;

    my %params;
    for my $key (qw/years months weeks days hours minutes seconds nanoseconds/) {
        $params{$key} = exists $args{$key} ? delete $args{$key} : 0;
    }
    if (%args) {
        my $msg = 'Invalid args: '.join ',', keys %args;
        Carp::croak $msg;
    }

    my $self = bless {
        months      => $params{months} + $params{years} * 12,
        days        => $params{days} + $params{weeks} * 7,
        minutes     => $params{minutes} + $params{hours} * 60,
        seconds     => $params{seconds},
        nanoseconds => $params{nanoseconds},
    } => $class;
    return $self->_normalize_nanoseconds();
}

# make the signs of seconds, nanos the same; 0 < abs(nanos) < MAX_NANOS
# NB this requires nanoseconds != 0 (callers check this already)
sub _normalize_nanoseconds {
    my $self = shift;
    return $self unless $self->{nanoseconds};

    my $seconds = $self->{seconds} + $self->{nanoseconds} / 1_000_000_000;
    $self->{seconds}     = int($seconds);
    $self->{nanoseconds} = $self->{nanoseconds} % 1_000_000_000;
    $self->{nanoseconds} -= 1_000_000_000 if $seconds < 0;

    return $self;
}

sub clone { bless {%{$_[0]}} => ref $_[0] }

sub years       { abs(shift->in_units(qw/years/))               }
sub months      { abs(shift->in_units(qw/months years/))        }
sub weeks       { abs(shift->in_units(qw/weeks/))               }
sub days        { abs(shift->in_units(qw/days weeks/))          }
sub hours       { abs(shift->in_units(qw/hours/))               }
sub minutes     { abs(shift->in_units(qw/minutes hours/) )      }
sub seconds     { abs(shift->in_units(qw/seconds/))             }
sub nanoseconds { abs(shift->in_units(qw/nanoseconds seconds/)) }

sub is_positive   { $_[0]->_has_positive  && !$_[0]->_has_negative }
sub is_negative   { !$_[0]->_has_positive && $_[0]->_has_negative }
sub _has_positive { (first { $_ > 0 } values %{$_[0]}) ? 1 : 0 }
sub _has_negative { (first { $_ < 0 } values %{$_[0]}) ? 1 : 0 }

sub is_zero {
    my $self = shift;
    return 0 if first { $_ != 0 } values %$self;
    return 1;
}

sub deltas { %{$_[0]} }

sub delta_months      { shift->{months}      }
sub delta_days        { shift->{days}        }
sub delta_minutes     { shift->{minutes}     }
sub delta_seconds     { shift->{seconds}     }
sub delta_nanoseconds { shift->{nanoseconds} }

sub in_units {
    my $self  = shift;
    my @units = @_;

    my %units = map { $_ => 1 } @units;

    my %ret;

    my ($months, $days, $minutes, $seconds) = @$self{qw/months days minutes seconds/};
    if ($units{years}) {
        $ret{years} = int($months / 12);
        $months -= $ret{years} * 12;
    }

    if ($units{months}) {
        $ret{months} = $months;
    }

    if ($units{weeks}) {
        $ret{weeks} = int($days / 7);
        $days -= $ret{weeks} * 7;
    }

    if ($units{days}) {
        $ret{days} = $days;
    }

    if ($units{hours}) {
        $ret{hours} = int($minutes / 60);
        $minutes -= $ret{hours} * 60;
    }

    if ($units{minutes}) {
        $ret{minutes} = $minutes;
    }

    if ($units{seconds}) {
        $ret{seconds} = $seconds;
        $seconds = 0;
    }

    if ($units{nanoseconds}) {
        $ret{nanoseconds} = $seconds * 1_000_000_000 + $self->{nanoseconds};
    }

    return wantarray ? @ret{@units} : $ret{$units[0]};
}

# XXX: limit mode only
sub is_wrap_mode      { 0 }
sub is_limit_mode     { 1 }
sub is_preserve_mode  { 0 }
sub end_of_month_mode { 'limit' }

sub calendar_duration {
    my $self = shift;
    my $clone = $self->clone;
    $clone->{$_} = 0 for qw/minutes seconds nanoseconds/;
    return $clone;
}

sub clock_duration {
    my $self = shift;
    my $clone = $self->clone;
    $clone->{$_} = 0 for qw/months days/;
    return $clone;
}

sub inverse {
    my $self = shift;
    my $clone = $self->clone;
    $clone->{$_} *= -1 for keys %$clone;
    return $clone;
}

sub add_duration {
    my ($lhs, $rhs) = @_;
    $lhs->{$_} += $rhs->{$_} for ALL_UNITS;
    return $lhs->_normalize_nanoseconds();
}

sub add {
    my $self = shift;
    my $class = ref $self;

    my $lhs = $self;
    my $rhs = $class->new(@_);
    return $lhs->add_duration($rhs);
}

sub subtract_duration { $_[0]->add_duration($_[1]->inverse) }

sub subtract {
    my $self = shift;
    my $class = ref $self;

    my $lhs = $self;
    my $rhs = $class->new(@_);
    return $lhs->subtract_duration($rhs);
}

sub multiply {
    my ($lhs, $rhs) = @_;
    $lhs->{$_} *= $rhs for ALL_UNITS;
    return $lhs->_normalize_nanoseconds();
}

sub compare {
    my ($class, $lhs, $rhs, $base) = @_;
    $base ||= DateTime::Moment->now;
    return DateTime::Moment->compare(
        $base->clone->add_duration($lhs),
        $base->clone->add_duration($rhs)
    );
}

sub _isa_datetime { blessed $_[0] && $_[0]->isa('DateTime') }

sub _add_overload {
    my ($lhs, $rhs, $flip) = @_;
    ($lhs, $rhs) = ($rhs, $lhs) if $flip;

    if (_isa_datetime($rhs)) {
        $rhs->add_duration($lhs);
        return;
    }

    # will also work if $lhs is a DateTime.pm object
    return $lhs->clone->add_duration($rhs);
}

sub _subtract_overload {
    my ($lhs, $rhs, $flip) = @_;
    ($lhs, $rhs) = ($rhs, $lhs) if $flip;

    if (_isa_datetime($rhs)) {
        Carp::croak('Cannot subtract a DateTime::Moment object from a DateTime::Moment::Duration object');
    }

    return $lhs->clone->subtract_duration($rhs);
}

sub _multiply_overload {
    my ($lhs, $rhs) = @_;
    return $lhs->clone->multiply($rhs);
}

sub _compare_overload {
    Carp::croak('DateTime::Moment::Duration does not overload comparison.  See the documentation on the compare() method for details.');
}

1;
__END__

=pod

=encoding utf-8

=head1 NAME

DateTime::Moment::Duration - TODO

=head1 SYNOPSIS

    use DateTime::Moment::Duration;

=head1 DESCRIPTION

TODO

=head1 SEE ALSO

L<DateTime::Duration>

=head1 LICENSE

Copyright (C) karupanerura.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

karupanerura E<lt>karupa@cpan.orgE<gt>

=cut


author/benchmark.pl
use strict;
use warnings;
use utf8;
use feature qw/say/;

use constant epoch => time;

use Benchmark qw/cmpthese timethese/;

use DateTime;
use DateTime::Moment;

sub hr { say '-' x 40 }

say 'new()';
cmpthese timethese 100000 => {
    datetime => sub { DateTime->new(year => 2016) },
    moment   => sub { DateTime::Moment->new(year => 2016) },
};
hr();

say 'now()';
cmpthese timethese 100000 => {
    datetime => sub { DateTime->now },
    moment   => sub { DateTime::Moment->now },
};
hr();

say 'from_epoch()';
cmpthese timethese 100000 => {
    datetime => sub { DateTime->from_epoch(epoch => epoch) },
    moment   => sub { DateTime::Moment->from_epoch(epoch => epoch) },
};
hr();

say 'calculate()';
cmpthese timethese 100000 => {
    datetime => sub { DateTime->now->add(years => 1) },
    moment   => sub { DateTime::Moment->now->add(years => 1) },
};
hr();

__END__
new()
Benchmark: timing 100000 iterations of datetime, moment...
  datetime:  4 wallclock secs ( 4.06 usr +  0.01 sys =  4.07 CPU) @ 24570.02/s (n=100000)
    moment:  1 wallclock secs ( 0.62 usr +  0.01 sys =  0.63 CPU) @ 158730.16/s (n=100000)
             Rate datetime   moment
datetime  24570/s       --     -85%
moment   158730/s     546%       --
----------------------------------------
now()
Benchmark: timing 100000 iterations of datetime, moment...
  datetime:  4 wallclock secs ( 4.38 usr +  0.01 sys =  4.39 CPU) @ 22779.04/s (n=100000)
    moment:  1 wallclock secs ( 0.59 usr +  0.00 sys =  0.59 CPU) @ 169491.53/s (n=100000)
             Rate datetime   moment
datetime  22779/s       --     -87%
moment   169492/s     644%       --
----------------------------------------
from_epoch()
Benchmark: timing 100000 iterations of datetime, moment...
  datetime:  4 wallclock secs ( 4.27 usr +  0.01 sys =  4.28 CPU) @ 23364.49/s (n=100000)
    moment:  1 wallclock secs ( 0.63 usr +  0.00 sys =  0.63 CPU) @ 158730.16/s (n=100000)
             Rate datetime   moment
datetime  23364/s       --     -85%
moment   158730/s     579%       --
----------------------------------------
calculate()
Benchmark: timing 100000 iterations of datetime, moment...
  datetime: 20 wallclock secs (20.30 usr +  0.04 sys = 20.34 CPU) @ 4916.42/s (n=100000)
    moment:  1 wallclock secs ( 1.07 usr +  0.00 sys =  1.07 CPU) @ 93457.94/s (n=100000)
            Rate datetime   moment
datetime  4916/s       --     -95%
moment   93458/s    1801%       --
----------------------------------------
